authappliance/menu.py

```diff
@@ -301,6 +303,140 @@ def _execute_remote_sql(self, sql):
         output = stdout.read()
         return output, err
 
+    def setup_tinc(self, local_vpn_ip, remote_vpn_ip, vpn_subnet, vpn_name='privacyideaVPN'):
+        """
+        Set up a tinc tunnel between self.local_ip and self.remote_ip
+        and update self.local_ip and self.remote_ip accordingly.
+        """
+        # create SFTP client to remote server
+        ssh = SSHClient()
+        ssh.set_missing_host_key_policy(AutoAddPolicy())
+        ssh.connect(str(self.remote_ip), username="root", password=self.password)
+        sftp = ssh.open_sftp()
+        local_io_handler = LocalIOHandler()
+        remote_io_handler = SFTPIOHandler(sftp)
+
+        tinc_net_directory = '/etc/tinc/{}/'.format(vpn_name)
+        # create directories locally and remotely
+        for io_handler in (local_io_handler, remote_io_handler):
+            io_handler.makedirs(os.path.join(tinc_net_directory, 'hosts'))
+        tinc_conf = os.path.join(tinc_net_directory, 'tinc.conf')
+        # Local tinc.conf
+        local_tinc_conf = TincConfFile(local_io_handler, tinc_conf)
+        local_tinc_conf.update({
+            'Name': 'pinode1',
+            'Device': '/dev/net/tun',
+            'AddressFamily': 'ipv4'
+        })
+        local_tinc_conf.save()
+
+        # Remote tinc.conf
+        remote_tinc_conf = TincConfFile(remote_io_handler, tinc_conf)
+        remote_tinc_conf.update({
+            'Name': 'pinode2',
+            'Device': '/dev/net/tun',
+            'AddressFamily': 'ipv4',
+            'ConnectTo': 'pinode1'
+        })
+        remote_tinc_conf.save()
+
+        # this command generates /etc/tinc/[vpn_name]/rsa_key.priv and /etc/tinc/[vpn_name]/hosts/[hostname]
+        generate_key_command = 'tincd -n {} -K 4096'.format(pipes.quote(vpn_name))
```

Comment:
Why do we use pipes?

Reply:
This is just a precaution to avoid shell injection. ``pipes`` is a Python stdlib module, so this doesn't create a new dependency.